### PR TITLE
python-bareos: use TLS-PSK from core ssl module (available since Python >= 3.13)

### PR DESCRIPTION
**Pull request by joergsteffens**
The ssl module in Python 3.13 (cpython) added support for TLS-PSK. This changes uses TLS-PSK from this core module, if the functionality is available. If not, it still fails back to sslpsk and than to unencrypted.

### Thank you for contributing to the Bareos Project!

**Backport of PR #0000 to bareos-2x** (remove this line if this is no backport; for backport use cherry-pick -x)

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Check backport line
- [x] Required backport PRs have been created

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

##### Tests
- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

File: python-bareos/bareos/bsock/lowlevel.py [hunk 1]
```diff
@@ -46,17 +46,21 @@
...
+# with implement TLS-PSK on top of the ssl module.
+# If it is also not available, we continue anyway,
 # but don't use TLS-PSK.
-try:
-    import sslpsk
-except ImportError:
-    warnings.warn(
-        "Connection encryption via TLS-PSK is not available, as the module sslpsk is not installed."
-    )
+if not getattr(ssl, "HAS_PSK", False):
```
Comment (bruno-at-bareos) [inline python-bareos/bareos/bsock/lowlevel.py:56]: Shouldn't we use the positive formulation
if getattr(ssl, "HAS_PSK", False):
    import ....
like it is used line 276, the not here may lead to confusion

File: python-bareos/bareos/bsock/lowlevel.py [hunk 2]
```diff
@@ -46,17 +46,21 @@
...
+    try:
+        import sslpsk
+    except ImportError:
+        warnings.warn(
+            "Connection encryption via TLS-PSK is not available "
+            "(TLS-PSK is not available in the ssl module and the extra module sslpsk is not installed)."
```
Comment (bruno-at-bareos) [inline python-bareos/bareos/bsock/lowlevel.py:62]: I found the warning too long on terminal output.

**Review by bruno-at-bareos — COMMENTED** [review]
Nice work, please see comments

File: python-bareos/bareos/bsock/lowlevel.py (hunk 1 above)
Comment (joergsteffens) [reply python-bareos/bareos/bsock/lowlevel.py:56]: I'm not sure, how to solve this. The only thing I can think of is:

```
if getattr(ssl, "HAS_PSK", False):
  # The ssl module has support for TLS-PSK.
  pass
else:
  # try to load sslpsk, to add TLS-PSK support to ssl.
  try:
    import sslpsk
    ...
```

Do you think, this is better? I'm not convinced.

**Comment by bruno-at-bareos** [comment]
One more thing we would like to treat as next LTS python will be not compatible with `sslpsk` library

with python >= 3.10 you can install sslpsk but it will failed when tried to used.

We certainly want after the first import do another try except block to catch this missing property, and then failed due to incompatible python/sslpsk mix.


```
======================================================================
ERROR: test_login_tls_tls_require (test_tlspsk.PythonBareosTlsPskTest.test_login_tls_tls_require)
console: tls, director: tls     => login
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/bareos/git/build/systemtests/tests/python-bareos/test_tlspsk.py", line 225, in test_login_tls_tls_require
    director = bareos.bsock.DirectorConsole(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/bareos/git/b-at-bareos/python-bareos/bareos/bsock/directorconsole.py", line 200, in __init__
    self.connect(
  File "/bareos/git/b-at-bareos/python-bareos/bareos/bsock/lowlevel.py", line 180, in connect
    return self.__connect()
           ^^^^^^^^^^^^^^^^
  File "/bareos/git/b-at-bareos/python-bareos/bareos/bsock/lowlevel.py", line 198, in __connect
    self.__connect_tls_psk()
  File "/bareos/git/b-at-bareos/python-bareos/bareos/bsock/lowlevel.py", line 284, in __connect_tls_psk
    self.socket = sslpsk.wrap_socket(
                  ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/site-packages/sslpsk/sslpsk.py", line 106, in wrap_socket
    _ssl_set_psk_client_callback(sock, cb)
  File "/usr/lib64/python3.11/site-packages/sslpsk/sslpsk.py", line 73, in _ssl_set_psk_client_callback
    ssl_id = _sslpsk.sslpsk_set_psk_client_callback(_sslobj(sock))
                                                    ^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/site-packages/sslpsk/sslpsk.py", line 55, in _sslobj
    return sock._sslobj._sslobj
           ^^^^^^^^^^^^^^^^^^^^
AttributeError: '_ssl._SSLSocket' object has no attribute '_sslobj'

```

File: python-bareos/bareos/bsock/lowlevel.py (hunk 2 above)
Comment (joergsteffens) [reply python-bareos/bareos/bsock/lowlevel.py]: I've shorted the warning.

**Comment by joergsteffens** [comment]
> One more thing we would like to treat as next LTS python will be not compatible with `sslpsk` library
> 
> with python >= 3.10 you can install sslpsk but it will failed when tried to used.
> 
> We certainly want after the first import do another try except block to catch this missing property, and then failed due to incompatible python/sslpsk mix.
> 
> ```
> ...
> AttributeError: '_ssl._SSLSocket' object has no attribute '_sslobj'
> ```

You mean an extra check in the tests?

**Comment by joergsteffens** [comment]
> > One more thing we would like to treat as next LTS python will be not compatible with `sslpsk` library
> > with python >= 3.10 you can install sslpsk but it will failed when tried to used.
> > We certainly want after the first import do another try except block to catch this missing property, and then failed due to incompatible python/sslpsk mix.
> > ```
> > ...
> > AttributeError: '_ssl._SSLSocket' object has no attribute '_sslobj'
> > ```
> 
> You mean an extra check in the tests?

I'm not sure how to cover this. The combination is broken, so the test should fail. I can do `try/except AttributeError` block around the connect, but still I should raise the problem after it has occurred. It seams, there exists a PR, that would solve these issues (https://github.com/drbild/sslpsk/pull/28), but it seams, as it will not get merged.

**Review by bruno-at-bareos — APPROVED** [review]
While there's certainly room for improvement, we want to merge this as it is now.
If someone decide to sponsor development around sslpsk forks like sslpsk3 we may then have the resources to even make it better.

**Comment by bruno-at-bareos** [comment]
So to resume the situation, for those who tried to install sslpsk from pypi and got the following error
```
AttributeError: '_ssl._SSLSocket' object has no attribute '_sslobj'
```
You have to remove and install sslpsk from github source.

**Pull request merged**